OpenTracing to OpenTelemetry Tutorial (#36789)
quarkusio/quarkus: docs/src/main/asciidoc/telemetry-opentracing-to-otel-tutorial.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
[id="telemetry-opentracing-to-otel-tutorial"]
= Migrate from OpenTracing to OpenTelemetry tracing
:categories: observability
:diataxis-type: tutorial
include::_attributes.adoc[]
:topics: observability,opentracing,opentelemetry,tracing,migration
:extensions: io.quarkus:quarkus-smallrye-opentracing,io.quarkus:quarkus-opentelemetry

Migrate an application from xref:opentracing.adoc.adoc[OpenTracing] to xref:opentelemetry.adoc[OpenTelemetry tracing] in Quarkus 3.x.

The legacy OpenTracing framework has been deprecated in favor of the new OpenTelemetry tracing framework. We announced the https://quarkus.io/blog/quarkus-observability-roadmap-2023/#opentracing-archived[OpenTracing deprecation on November 2022], and we are dropping the extension from core Quarkus and move it to the Quarkiverse.

It is now time to migrate your application to OpenTelemetry tracing if you haven’t done it yet.

> If you need to migrate to Quarkus 2.16.x please beware that configuration properties are different and you should check the olded Quarkus OpenTelemetry guide version, https://quarkus.io/version/2.16/guides/opentelemetry#configuration-reference[here].

Reviewer: should it not be older instead of olded?

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
[id="telemetry-opentracing-to-otel-tutorial"]
= Migrate from OpenTracing to OpenTelemetry tracing
:categories: observability
:diataxis-type: tutorial
include::_attributes.adoc[]
:topics: observability,opentracing,opentelemetry,tracing,migration
:extensions: io.quarkus:quarkus-smallrye-opentracing,io.quarkus:quarkus-opentelemetry

> Migrate an application from xref:opentracing.adoc.adoc[OpenTracing] to xref:opentelemetry.adoc[OpenTelemetry tracing] in Quarkus 3.x.

Author: yes, after the migration to quarkiverse we update the link.

It is now time to migrate your application to OpenTelemetry tracing if you haven’t done it yet.

If you need to migrate to Quarkus 2.16.x please beware that configuration properties are different and you should check the older Quarkus OpenTelemetry guide version, https://quarkus.io/version/2.16/guides/opentelemetry#configuration-reference[here].

== Prerequisites

include::{includes}/prerequisites.adoc[]

== Summary

The demo has 5 parts. Please read the summary and then jump to the section that best fits your use case.

1 - The *starting point* presents the quickstart app that uses OpenTracing

2 - The first part is good for anyone performing a *big bang change* of OpenTracing when you don't have any manual instrumentation

3 - This is the *big bang replacement* of OpenTracing when you have manually instrumented the code. We explain the main differences between OpenTracing and OpenTelemetry

> 4 - The last part uses the *OpenTracing shim*. This is useful if you have a large application with manually instrumented the code. It can help performing the migration step by step because it allows the use of the legacy OpenTracing API on top of new OpenTelemetry API

Author: The sentence seems clunky. With rewrite.

5 - Conclusion and additional resources

The tasks described bellow fall into 3 categories:

* Dependencies
* Configuration
* Code

[[starting-point]]
== Starting point

> The base project was created with Quarkus code generator. At some point that generator will not be able to create projects by using the legacy `quarkus-smallrye-opentracing` extension anymore. For that reason, you can checkout this demo project from GitHub:

Author: Will fix.

=== Checkout the existing legacy project

[source,bash]
----

> git clone git@github.com:brunobat/opentracing-quickstart.git

Author: Where should I place it?

> git clone git@github.com:brunobat/opentracing-quickstart.git

Reviewer 2: Can you initiate a move to the quarkusio org? I think it would be better this way. I let you do that and adjust the URL accordingly and we should be good.

FYI, I moved the Jaeger and SmallRye OpenTracing extensions to the Quarkiverse. I will wait for 3.6.0 core to be released to release them so that they only appear for 3.6+.

I will proceed with the PR to remove them as soon as this is in as I will need to adjust your guide in this PR (the opentracing.adoc doc link).

> git clone git@github.com:brunobat/opentracing-quickstart.git

Author: Done

----

For convenience, https://github.com/brunobat/opentracing-quickstart[the repository] containing the app to migrate, includes several branches with commits mimicking the migration steps described in this guide. You can checkout the `main` branch to start from the beginning.

=== Generate the legacy project

Create the legacy project by using the https://code.quarkus.io/[Quarkus code generator] with the `resteasy-reactive` and `quarkus-smallrye-opentracing` extensions, or by executing the following command:

:create-app-artifact-id: opentracing-quickstart
:create-app-extensions: resteasy-reactive,quarkus-smallrye-opentracing
:create-app-code:
include::{includes}/devtools/create-app.adoc[]

This command generates the Maven project and imports the `smallrye-opentracing` extension, which

> includes the OpenTracing support and the default https://www.jaegertracing.io/[Jaeger] tracer.

Reviewer 2: Could you try to be a bit more consistent with the rest of the guides here? Because typically you will also have the Gradle instructions there.

You don't need the full verbiage here (typically I don't think you need the POM/build.gradle content) but being more consistent would help.

Also, we usually don't mention the code.quarkus.io alternative.

See https://quarkus.io/guides/validation#creating-the-maven-project

Author: done

=== The application

> The Quarkus project has a single endpoint and the related class will look like this:

Reviewer 2: ```suggestion
The Quarkus project has a single endpoint and the related class looks like this:
```

        return "Hello from RESTEasy Reactive";
    }
}
----

There is no OpenTracing specific code in the generated project, but the `smallrye-opentracing` extension is present and enabled by default, and it will automatically instrument the code.

Let's start the Jaeger-all-in-one Docker image, where we will retrieve and see the captured traces:

[source,bash]
----
docker run -e COLLECTOR_OTLP_ENABLED=true -p 6831:6831/udp -p 6832:6832/udp -p 5778:5778 -p 16686:16686 -p 4317:4317 -p 4318:4318 -p 14250:14250 -p 14268:14268 -p 14269:14269 -p 9411:9411 jaegertracing/all-in-one:latest
----

At this point you can run the application with Quarkus dev mode:

include::{includes}/devtools/dev.adoc[]

If you call the http://localhost:8080/hello[`/hello` endpoint] the related traces can be retrieved in the Jaeger UI at this address: http://localhost:16686/

They will look like this:

image::ot-to-otel-1.png[alt=OpenTracing span,role="center"]

== Big bang change from OpenTracing to OpenTelemetry

This is the happiest path, in this case there is no manual instrumentation. We can do a big bang change from OpenTracing to OpenTelemetry without side effects.

=== Change dependencies

> To migrate between the two frameworks, you must drop the old `quarkus-smallrye-opentracing` extension and replace it by the new the `quarkus-opentelemetry` extension in the pom.xml file:

Reviewer 2: ```suggestion
To migrate between the two frameworks, you must drop the old `quarkus-smallrye-opentracing` extension and replace it by the `quarkus-opentelemetry` extension in the build file:
```

The legacy extension is removed from the project:
[source,xml]
----
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-smallrye-opentracing</artifactId>
</dependency>
----

The new one is added:
[source,xml]
----
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-opentelemetry</artifactId>
</dependency>

> ----

Reviewer 2: Could you use the tabs here? See example here: https://github.com/quarkusio/quarkus/blob/main/docs/src/main/asciidoc/validation.adoc?plain=1#L57-L70

Author: done

=== Application properties

You should remove the old OpenTracing properties, starting with `quarkus.jaeger.*` from the `application.properties` file, like in this example:
[source,application.properties]
----
#Legacy OpenTracing properties to be removed
quarkus.jaeger.service-name=legume
quarkus.jaeger.sampler-type=const
quarkus.jaeger.sampler-param=1
quarkus.jaeger.endpoint=http://localhost:14268/api/traces
quarkus.jaeger.log-trace-context=true
----

> If you use the default values in the OpenTelemetry properties, there is no necessity to include anything in the `application.properties` file.

Reviewer 2: Add a new line here after the `----`. You never know how the other publishing tool will react.

Author: ok

Some common properties to migrate are:

|===
|Legacy OpenTracing property | New OpenTelemetry property

|`quarkus.jaeger.service-name=legume`
|`quarkus.application.name=legume`

|`quarkus.jaeger.endpoint=http://localhost:14268/api/traces`
|`quarkus.otel.exporter.otlp.traces.endpoint=http://localhost:4317`

|`quarkus.jaeger.auth-token`
|`quarkus.otel.exporter.otlp.traces.headers`

|`quarkus.jaeger.sampler-type`
|`quarkus.otel.traces.sampler`

|`quarkus.jaeger.sampler-param`
|`quarkus.otel.traces.sampler.arg`

|`quarkus.jaeger.tags`
|`quarkus.otel.resource.attributes`

|`quarkus.jaeger.propagation`
|`quarkus.otel.propagators`
|===

The way the extensions can be enabled and disabled is very different. The OpenTelemetry extension is enabled by default and you can disable all or parts of it by checking xref:opentelemetry.adoc#disable-all-or-parts-of-the-opentelemetry-extension[this section of the OpenTelemetry guide].

> All the OpenTelemetry properties and their defaults can be found at the xref:opentelemetry.adoc#configuration-reference[OpenTelemetry configuration].

Reviewer 2: ```suggestion
All the OpenTelemetry properties and their defaults can be found in the xref:opentelemetry.adoc#configuration-reference[OpenTelemetry configuration reference].
```

=== Run the application

Restarting Quarkus is not needed, auto-reload should have kicked in and you now can call the http://localhost:8080/hello[`/hello` endpoint] and then see the traces in the Jaeger UI: http://localhost:16686/

However, you can now see spans produced by the OpenTelemetry's auto-instrumentation instead of the OpenTracing one:

image::ot-to-otel-2.png[alt=OpenTelemetry span,role="center"]

If you don't have any manual instrumentation of your own, you are done!

== The big bang replacement, when you have manual instrumentation

Let's say instead of the `GreetingResource` class from above, you have something more complex. You will need additional work on top of the changes from the <<starting-point>>.

> This class will now use the `@Traced` annotation and creates a "manual" programmatic span.

Reviewer 2: ```suggestion
This class now uses the `@Traced` annotation and creates a "manual" programmatic span.
```

> Please copy/paste that code for the `GreetingResource` class in the quickstart project:

Reviewer 2: ```suggestion
Copy/paste that code for the `GreetingResource` class in the quickstart project:
```

import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

import static io.opentelemetry.api.trace.StatusCode.*;

@Path("/hello")
@ApplicationScoped
public class GreetingResource {

    @Inject
    io.opentelemetry.api.trace.Tracer otelTracer;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @WithSpan(value = "Not needed, will create a new span, child of the automatic JAX-RS span")
    public String hello() {
        // Add a tag to the active span
        Span incomingSpan = Span.current();
        incomingSpan.setAttribute(SemanticAttributes.CODE_NAMESPACE, "GreetingResource");

        // Create a manual inner span
        Span innerSpan = otelTracer.spanBuilder("Count response chars").startSpan();
        try (Scope scope = innerSpan.makeCurrent()) {
            final String response = "Hello from RESTEasy Reactive";
            innerSpan.setAttribute("response-chars-count", response.length());
            return response;
        } catch (Exception e) {
            innerSpan.setStatus(ERROR);
            innerSpan.recordException(e);
            throw e;
        } finally {
            innerSpan.end();
        }
    }
}

----

> Once you remove all the OpenTracing dependencies the code will build. Don't forget to double check if the traces contain the right spans. You can see them in the Jaeger UI: http://localhost:16686/.

Reviewer 2: Same here, new line after the `----`

Author: ok

== The OpenTracing shim

> On this part we present the OpenTelemetry library that can smooth the transition by providing access to the legacy OpenTracing API. This can help with the migration of large applications with many manual instrumentation points.

Reviewer 2: ```suggestion
In this section, we present an OpenTelemetry library that can smooth the transition by providing access to the legacy OpenTracing API. This can help with the migration of large applications with many manual instrumentation points.
```

=== The dependencies

> Remove the `quarkus-smallrye-opentracing` extension and add the `quarkus-opentelemetry` extension and the`opentelemetry-opentracing-shim` library to the `pom.xml` file:

Reviewer 2: ```suggestion
Remove the `quarkus-smallrye-opentracing` extension and add the `quarkus-opentelemetry` extension and the `opentelemetry-opentracing-shim` library to the build file:
```

> Remove the `quarkus-smallrye-opentracing` extension and add the `quarkus-opentelemetry` extension and the`opentelemetry-opentracing-shim` library to the `pom.xml` file:

Reviewer 2: I think you should make it clear somehow that it is not a followup from the previous section as I suspect it won't work right? Maybe by saying that you should start from the unmodified quickstart?

Author: will rewrite

The legacy extension is removed from the project:
[source,xml]
----
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-smallrye-opentracing</artifactId>
</dependency>
----

The new one is added:
[source,xml]
----
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-opentelemetry</artifactId>
</dependency>
<dependency>
  <groupId>io.opentelemetry</groupId>
  <artifactId>opentelemetry-opentracing-shim</artifactId>
  <!-- No need to declare the version -->
</dependency>

> ----

Reviewer 2: Please use the tabs here too.

Author: ok

import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import org.eclipse.microprofile.opentracing.Traced;

@Path("/hello")
@ApplicationScoped
public class GreetingResource {

    @Inject
    io.opentracing.Tracer legacyTracer; <1>

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @Traced(operationName = "Not needed, will change the current span name") <2>
    public String hello() {
        // Add a tag to the active span
        legacyTracer.activeSpan().setTag(Tags.COMPONENT, "GreetingResource"); <3>

        // Create a manual inner span
        Span innerSpan = legacyTracer.buildSpan("Count response chars").start();

        try (Scope dbScope = legacyTracer.scopeManager().activate(innerSpan)) {
            String response = "Hello from RESTEasy Reactive";
            innerSpan.setTag("response-chars-count", response.length());
            return response;
        } catch (Exception e) {
            innerSpan.setTag("error", true);
            innerSpan.setTag("error.message", e.getMessage());
            throw e;
        } finally {
            innerSpan.finish();
        }
    }
}
----

> <1> The `Tracer` annotation must be removed and on its place, we need to inject the OpenTelemetry SDK. We will need it in <7>.

Reviewer 2: ```suggestion
<1> The `Tracer` annotation must be removed and instead, we need to inject the OpenTelemetry SDK. We will need it in <7>.
```

> <1> The `Tracer` annotation must be removed and on its place, we need to inject the OpenTelemetry SDK. We will need it in <7>.

Reviewer 2: Not sure what `<7>` is? I don't see it.

Author: yeah... it's a leftover. Will fix.
